File: workspace_agent/file_tools/image_tools.py
"""
画像ファイル用ツール（コンテナ側）

inspect_image_file: メタデータ確認（解像度、サイズ）
※ read_image_file は registry.py に実装
"""

import io
from typing import Any

import structlog

from workspace_agent.file_tools.utils import (
    local_file_tool_handler,
    format_tool_error,
    format_tool_success,
)

logger = structlog.get_logger(__name__)

# 画像としてサポートするMIMEタイプ
IMAGE_MIME_TYPES = {
    "image/jpeg",
    "image/jpg",
    "image/png",
    "image/gif",
    "image/webp",
    "image/bmp",
    "image/tiff",
    "image/svg+xml",
}


def _is_image_file(filename: str, content_type: str | None) -> bool:
    """ファイルが画像かどうかを判定"""
    if content_type and content_type in IMAGE_MIME_TYPES:
        return True

    # 拡張子でフォールバック判定
    lower_name = filename.lower()
    image_extensions = {".jpg", ".jpeg", ".png", ".gif", ".webp", ".bmp", ".tiff", ".tif", ".svg"}
    for ext in image_extensions:
        if lower_name.endswith(ext):
            return True

    return False


@local_file_tool_handler(log_prefix="画像情報取得")
async def inspect_image_file_handler(*, content, filename, content_type, args, **_):
    """
    画像ファイルのメタデータを確認

    Args:
        args:
            file_path: ファイルパス
    """
    # 画像ファイルかチェック
    if not _is_image_file(filename, content_type):
        return format_tool_error(
            f"このファイルは画像ではありません: {filename} ({content_type})"
        )

    result_lines = [
        f"# 画像情報: {filename}",
        f"ファイルサイズ: {len(content) / 1024:.1f} KB",
        f"MIMEタイプ: {content_type}",
    ]

    # Pillowで詳細情報を取得
    try:
        # オプション依存: 未インストール時はフォールバック
        from PIL import Image
        from PIL.ExifTags import TAGS

        img = Image.open(io.BytesIO(content))

        result_lines.append(f"解像度: {img.width} x {img.height} px")
        result_lines.append(f"カラーモード: {img.mode}")
        result_lines.append(f"フォーマット: {img.format}")

        # DPI情報
        dpi = img.info.get("dpi")
        if dpi:
            result_lines.append(f"DPI: {dpi[0]} x {dpi[1]}")

        # アニメーション（GIF）
        if hasattr(img, "n_frames") and img.n_frames > 1:
            result_lines.append(f"フレーム数: {img.n_frames}")

        # EXIF情報（簡易版）
        if hasattr(img, "_getexif") and img._getexif():
            exif = img._getexif()
            result_lines.append("")
            result_lines.append("## EXIF情報")

            important_tags = {
                "Make": "カメラメーカー",
                "Model": "カメラ機種",
                "DateTime": "撮影日時",
                "ExposureTime": "シャッタースピード",
                "FNumber": "絞り値",
                "ISOSpeedRatings": "ISO感度",
            }

            for tag_id, value in exif.items():
                tag_name = TAGS.get(tag_id, str(tag_id))
                if tag_name in important_tags:
                    result_lines.append(f"- {important_tags[tag_name]}: {value}")

        img.close()

    except ImportError:
        result_lines.append("")
        result_lines.append("※ Pillowがインストールされていないため、詳細情報を取得できません。")
    except Exception as e:
        logger.warning("画像詳細情報取得エラー", error=str(e))
        result_lines.append("")
        result_lines.append(f"※ 詳細情報の取得に失敗: {str(e)}")

    result_text = "\n".join(result_lines)
    result_text += "\n\n---\n"
    result_text += "画像を視覚的に確認するには `read_image_file` を使用してください。"

    return format_tool_success(result_text)

File: workspace-agent/src/file-tools/image-tools.ts
/**
 * Image file tools
 *
 * inspect_image_file: Metadata inspection (resolution, size)
 * Note: read_image_file is implemented in registry.ts
 */

import sharp from "sharp";
import { createLogger } from "../logger.js";
import {
  localFileToolHandler,
  formatToolError,
  formatToolSuccess,
  type ToolResult,
} from "./utils.js";

const logger = createLogger("image-tools");

// Supported image MIME types
const IMAGE_MIME_TYPES = new Set([
  "image/jpeg",
  "image/jpg",
  "image/png",
  "image/gif",
  "image/webp",
  "image/bmp",
  "image/tiff",
  "image/svg+xml",
]);

// Supported image extensions
const IMAGE_EXTENSIONS = new Set([
  ".jpg",
  ".jpeg",
  ".png",
  ".gif",
  ".webp",
  ".bmp",
  ".tiff",
  ".tif",
  ".svg",
]);

/**
 * Check if a file is an image based on content type or extension
 */
export function isImageFile(
  filename: string,
  contentType: string | null,
): boolean {
  if (contentType && IMAGE_MIME_TYPES.has(contentType)) {
    return true;
  }

  // Fallback to extension check
  const lowerName = filename.toLowerCase();
  for (const ext of IMAGE_EXTENSIONS) {
    if (lowerName.endsWith(ext)) {
      return true;
    }
  }

  return false;
}

/**
 * inspect_image_file handler - Get image file metadata
 */
export const inspectImageFileHandler = localFileToolHandler(
  { logPrefix: "画像情報取得" },
  async ({ content, filename, contentType }): Promise<ToolResult> => {
    // Check if it's an image file
    if (!isImageFile(filename, contentType)) {
      return formatToolError(
        `このファイルは画像ではありません: ${filename} (${contentType})`,
      );
    }

    const resultLines: string[] = [
      `# 画像情報: ${filename}`,
      `ファイルサイズ: ${(content.length / 1024).toFixed(1)} KB`,
      `MIMEタイプ: ${contentType}`,
    ];

    // Get detailed info using sharp
    try {
      const metadata = await sharp(content).metadata();

      if (metadata.width && metadata.height) {
        resultLines.push(`解像度: ${metadata.width} x ${metadata.height} px`);
      }
      if (metadata.space) {
        resultLines.push(`カラーモード: ${metadata.space}`);
      }
      if (metadata.format) {
        resultLines.push(`フォーマット: ${metadata.format}`);
      }
      if (metadata.density) {
        resultLines.push(`DPI: ${metadata.density}`);
      }
      if (metadata.pages && metadata.pages > 1) {
        resultLines.push(`フレーム数: ${metadata.pages}`);
      }
      if (metadata.hasAlpha !== undefined) {
        resultLines.push(
          `アルファチャンネル: ${metadata.hasAlpha ? "あり" : "なし"}`,
        );
      }

      // EXIF info (if available)
      if (metadata.exif) {
        try {
          // sharp provides raw EXIF buffer; we report its presence
          resultLines.push("");
          resultLines.push("## EXIF情報");
          resultLines.push("- EXIF データが含まれています");
          if (metadata.orientation) {
            resultLines.push(`- 回転情報: ${metadata.orientation}`);
          }
        } catch {
          // EXIF parsing failed, skip
        }
      }
    } catch (e) {
      logger.warn({ err: e }, "画像詳細情報取得エラー");
      resultLines.push("");
      resultLines.push(
        `※ 詳細情報の取得に失敗: ${e instanceof Error ? e.message : String(e)}`,
      );
    }

    let resultText = resultLines.join("\n");
    resultText += "\n\n---\n";
    resultText +=
      "画像を視覚的に確認するには `read_image_file` を使用してください。";

    return formatToolSuccess(resultText);
  },
);
